ccadb/ccadb.go

```diff
@@ -0,0 +1,289 @@
+				}
+				serials[serialByteString] = crl
+			}
+
+			age := time.Since(crl.ThisUpdate).Round(time.Minute)
+			nextUpdate := time.Until(crl.NextUpdate).Round(time.Hour)
+			entries += len(crl.RevokedCertificateEntries)
+			bytes += len(crl.Raw)
+			log.Printf("crl %q: %d entries, %d bytes, age %gm, nextUpdate %gh", url, len(crl.RevokedCertificateEntries), len(crl.Raw), age.Minutes(), nextUpdate.Hours())
+		}
+	}
+
+	log.Printf("%d CRLs had %d entries and %d bytes", crls, entries, bytes)
+	return errors.Join(errs...)
+}
+
+func checkCRL(ctx context.Context, url string, issuer *x509.Certificate, ageLimit time.Duration) (*x509.RevocationList, error) {
+	req, err := http.NewRequestWithContext(ctx, http.MethodGet, url, nil)
+	if err != nil {
+		return nil, err
+	}
+	resp, err := http.DefaultClient.Do(req)
+	if err != nil {
+		return nil, err
+	}
+	defer resp.Body.Close()
+	if resp.StatusCode != http.StatusOK {
+		return nil, fmt.Errorf("HTTP status code %d", resp.StatusCode)
+	}
+	body, err := io.ReadAll(resp.Body)
+	if err != nil {
+		return nil, fmt.Errorf("reading CRL body: %s", err)
+	}
+	crl, err := x509.ParseRevocationList(body)
+	if err != nil {
+		return nil, err
+	}
+
+	idp, err := idp.Get(crl)
+	if err != nil {
+		return nil, err
+	}
+
+	if idp != url {
+		return nil, fmt.Errorf("CRL fetched from %s had mismatched IDP %s", url, idp)
+	}
+
+	return crl, checker.Validate(crl, issuer, ageLimit)
+}
+
+// returns a map from issuer SKID to list of URLs
+func (c Checker) getCRLURLs(ctx context.Context, csvURL string, owner string) (map[string][]string, error) {
+	req, err := http.NewRequestWithContext(ctx, http.MethodGet, csvURL, nil)
+	if err != nil {
+		return nil, err
+	}
+	resp, err := http.DefaultClient.Do(req)
+	if err != nil {
+		return nil, err
+	}
+	defer resp.Body.Close()
+	if resp.StatusCode != http.StatusOK {
+		return nil, fmt.Errorf("HTTP status code %d", resp.StatusCode)
+	}
+	reader := csv.NewReader(resp.Body)
+	header, err := reader.Read()
+	if err != nil {
+		return nil, err
+	}
+
+	var ownerIndex, crlIndex, skidIndex, certificateNameIndex int
+	for i, name := range header {
+		if name == "CA Owner" {
+			ownerIndex = i
+		}
+		if name == "JSON Array of Partitioned CRLs" {
+			crlIndex = i
+		}
+		if name == "Subject Key Identifier" {
+			skidIndex = i
+		}
+		if name == "Certificate Name" {
+			certificateNameIndex = i
+		}
+	}
+	allCRLs := make(map[string][]string)
+	for {
+		record, err := reader.Read()
+		if err == io.EOF {
+			break
+		}
+		if err != nil {
+			return nil, err
+		}
+		if record[ownerIndex] != owner {
+			continue
+		}
+		crlJSON := record[crlIndex]
+		if crlJSON == "" {
+			continue
+		}
+		var crls []string
+		err = json.Unmarshal([]byte(crlJSON), &crls)
+		if err != nil {
+			return nil, err
+		}
+		// Roots have a CRL list containing a single ""
+		if len(crls) == 1 && crls[0] == "" {
+			continue
+		}
```

Comment:
If the idea is to skip root CRLs, then I would suggest we directly check the `Certificate Record Type` column to be `Root Certificate`. Checking it via heuristics makes me a bit nervous.

Reply:
Good call. I just deleted this check. I believe it was used in an earlier version, where I was matching entries on organization name, and our root certificates matched. Now I'm matching based on SKID matches against an embedded bundle that contains only our intermediates.